SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp

```diff
@@ -193,3 +196,114 @@ void USpatialNetConnection::OnHeartbeat()
 {
 	SetHeartbeatTimeoutTimer();
 }
+
+void USpatialNetConnection::InitPing(FTimerManager* InTimerManager, Worker_EntityId InPlayerControllerEntity)
+{
+	checkf(PlayerControllerEntity == SpatialConstants::INVALID_ENTITY_ID || PlayerControllerEntity == InPlayerControllerEntity, TEXT("InitPing: PlayerControllerEntity already set: %lld. New entity: %lld"), PlayerControllerEntity, InPlayerControllerEntity);
+	PlayerControllerEntity = InPlayerControllerEntity;
+	TimerManager = TimerManager == nullptr ? InTimerManager : TimerManager;
+
+	TWeakObjectPtr<USpatialNetConnection> ConnectionPtr = this;
+
+	if (Driver->IsServer())
+	{
+		Cast<USpatialNetDriver>(Driver)->Receiver->AddClientPongDelegate(PlayerControllerEntity, ClientPongDelegate::CreateLambda([ConnectionPtr](Worker_ComponentUpdateOp& Op)
+		{
+			if (ConnectionPtr.IsValid())
+			{
+				float fReceivedTimestamp = ConnectionPtr->GetWorld()->GetTimeSeconds();
+
+				Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(Op.update.schema_type);
+				uint32 EventCount = Schema_GetObjectCount(EventsObject, SpatialConstants::PING_EVENT_ID);
+
+				if (EventCount > 0 && ConnectionPtr->PlayerController)
+				{
+					if (APlayerState* PlayerState = Cast<APlayerController>(ConnectionPtr->PlayerController)->PlayerState)
+					{
+						uint32 ID = Schema_GetUint32(EventsObject, SpatialConstants::PING_ID_OFFSET_ID);
+						float fTimeSent;
+						if (ConnectionPtr->SentPingTimestamps.RemoveAndCopyValue(ID, fTimeSent))
+						{
+							// Set player state ExactPing in msecs, ExactPing is not replicated
+							PlayerState->ExactPing = (fReceivedTimestamp - fTimeSent) * 1000.0f;
+
+							// In native Unreal, Ping property on player state is replicated
+							// (compressed for replication by dividing ExactPing by 4)
```

Comment:
Presumably this is so we drop the 2 least major bits. Slightly faster and, IMO, explicit if we turn this into an int then divide. (If we know that this will never be negative then shift right might be clearer for what we are doing).

Reply:
Using `APlayerState::UpdatePing` to do ping calculation and compression